### ISSUE TITLE
Is this compatible with DeepSpeed / ZeRO?

### ISSUE THREAD

**Issue by StellaAthena** [microsoft/mup#6]
I train [large language models](https://github.com/EleutherAI/gpt-neox) using DeepSpeed's ZeRO optimizer. Does this library support ZeRO?

**Comment by edwardjhu** [comment]
Hi Stella,

Thanks for bringing our attention to the compatibility with DeepSpeed!

In principle, yes. Our muAdam is just a wrapper that scales the learning rates for different parameter groups differently before passing them on to Adam.

See https://github.com/microsoft/mup/blob/c9d67001c47ae254ea4b7e26146ffd059520b6ba/mup/optim.py#L38

If I understand correctly, ZeRO is a technique that makes the underlying optimizer more efficient without changing the math. If you could pass the parameter groups created by muAdam to ZeRO Adam, it should work as expected.

Please let us know if our understanding of ZeRO is correct. We'll also look into what the best way to integrate with DeepSpeed is, given its popularity among folks who train large models.

**Comment by zhuzilin** [comment]
@StellaAthena @edwardjhu Our team is also trying to migrate muP to gpt-neox. It seems to me that `MuAdam` works fine with zero optimizer, as all it does is adjusting the learning rate. Figure below is the "coord check" test for a 12 layer model (basically the `small.yml` config in gpt-neox), width from 12 to 768:
![image](https://user-images.githubusercontent.com/10428324/160589783-ad7680ad-9ec1-43f0-b58b-e21297f6578a.png)
As you can see, mup works well with gpt-neox!

I think the major barrier to use mup is that we could not use the `MuReadout` in mup as the output layer. We need to have a model-parallel version. And applying the coord check is hard on large model, because deepspeed will somehow occupy certain gpu memory even after the model is garbaged collected and current coord check impl need to initialized multiple model squentially, which will lead to OOM for multiple large model.

If you have interest in adapting mup to gpt-neox, maybe we can work together :)

**Comment by thegregyang** [comment]
@zhuzilin So `MuReadout` essentially just does the following
```python
    def forward(self, x):
        return super().forward(
            self.output_mult * x /  self.weight.infshape.width_mult())
```
If for some reason you can't use `MuReadout` as is, you can manually write out this `forward`, assuming you have already called `set_base_shapes`. For example, in your model `forward` function,
```python
output = x @ self.output_weights.T * self.output_mult / self.output_weights.infshape.width_mult() + self.output_bias
```
I'm not sure what you need exactly for model parallelism, but hopefully this helps. But if you do figure out how to do model parallelism, we would like to hear how to incorporate your insight into `mup` to make it work right out the box.

**Comment by StellaAthena** [comment]
@zhuzilin I'm looking into adding support for muP to GPT-NeoX and would love to check out your code. Where can I find it?

**Comment by thegregyang** [comment]
@zhuzilin @StellaAthena How is Deepspeed integration going? We can connect you with members of the Deepspeed team if necessary.
